### PR TITLE
csi: checkpoint volume claim garbage collection

### PR DESCRIPTION
**Pull request by tgross**
Part of fixing https://github.com/hashicorp/nomad/issues/7629
Pulled out of https://github.com/hashicorp/nomad/pull/7708

Adds a `CSIVolumeClaim` type to be tracked as current and past claims
on a volume. Allows for a client RPC failure during node or controller
detachment without having to keep the allocation around after the
first garbage collection eval.

This changeset lays groundwork for moving the actual detachment RPCs
into a volume watching loop outside the GC eval.

### CODE REVIEW

**Comment by tgross** [comment]
@lang I need to make sure this works end-to-end but it should be ready for an overall review.

**Comment by tgross** [comment]
I've tested this with both `nomad job stop` and `nomad job stop -purge` cases and I've got properly detached EBS volumes and released claims. Marking this ready for review.

File: nomad/structs/csi.go [hunk 1]
```diff
@@ -350,50 +374,65 @@ func (v *CSIVolume) Copy() *CSIVolume {
 		out.WriteAllocs[k] = v
 	}
 
+	for k, v := range v.ReadClaims {
```
Comment (langmartin) [inline nomad/structs/csi.go:377]: do we need to copy the claims themselves here too? the `state` is mutable

Comment (tgross) [reply nomad/structs/csi.go:377]: Good catch! Will fix!

Comment (tgross) [reply nomad/structs/csi.go:377]: Done

**Pull request merged**